src/Npgsql/NpgsqlNestedDataReader.cs

```diff
@@ -0,0 +1,509 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandlers;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.IO;
+using System.Runtime.CompilerServices;
+
+namespace Npgsql
+{
+    /// <summary>
+    /// Reads a forward-only stream of rows from a nested data source.
+    /// Can be retrieved using <see cref="NpgsqlDataReader.GetData(int)"/> or
+    /// <see cref="NpgsqlNestedDataReader.GetData(int)"/>.
+    /// </summary>
+    public sealed class NpgsqlNestedDataReader : DbDataReader
+    {
+        NpgsqlDataReader _outermostReader;
+        ulong _uniqueOutermostReaderRowId;
+        NpgsqlNestedDataReader? _outerNestedReader;
+        NpgsqlNestedDataReader? _cachedFreeNestedDataReader;
+        int _depth;
+        int _numRows;
+        int _nextRowIndex;
+        int _nextRowBufferPos;
+        ReaderState _readerState;
+
+        readonly List<ColumnInfo> _columns = new();
+
+        struct ColumnInfo
+        {
+            public uint TypeOid;
+            public int BufferPos;
+            public NpgsqlTypeHandler TypeHandler;
+        }
+
+        NpgsqlReadBuffer Buffer => _outermostReader.Buffer;
+        ConnectorTypeMapper TypeMapper => _outermostReader.Connector.TypeMapper;
+
+        internal NpgsqlNestedDataReader(NpgsqlDataReader outermostReader, ulong uniqueOutermostReaderRowId, int depth)
+        {
+            _outermostReader = outermostReader;
+            _uniqueOutermostReaderRowId = uniqueOutermostReaderRowId;
+            _depth = depth;
+        }
+
+        internal void Recreate(NpgsqlDataReader outermostReader, ulong uniqueOutermostReaderRowId, int depth)
+        {
+            _outermostReader = outermostReader;
```

Comment:
Do the outermost reader (and depth) ever change, if the nested reader is always cached on its parent reader? If so, they can be made readonly.